### PR TITLE
PR1 | Functions | Add structures to store function metadata

### PR DESCRIPTION
**Pull request by Sakthi-SM**
**What does this Pull Request accomplish?**

This PR defines basic structures to store the function metadata and these will be used in the template file to generate the necessary files.

**Why should this Pull Request be merged?**
This implements the [Task 2296082](https://dev.azure.com/ni/DevCentral/_workitems/edit/2296082): Add a data structure to contain all the information related to the functions that are to be generated

### CODE REVIEW

File: src/codegen/properties/functions.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from codegen.properties.parameter import Parameter
+
+
+class Functions:
+    """Structure for storing function metadata from scrapigen."""
+
+    def __init__(self, function_name, function_metadata):
+        """Structure for storing function metadata from scrapigen."""
+        self._function_name = function_name
+        self._c_function_name = function_metadata["c_function_name"]
```
Comment (bkeryan) [inline src/codegen/properties/functions.py:13]: `c_function_name` no longer exists.

```suggestion
        if "cname" in function_metadata:
            assert function_metadata["cname"].startswith("DAQmx")
            self._c_function_name = function_metadata["cname"][5:]
        else:
            self._c_function_name = function_name
```

File: src/codegen/properties/functions.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        self._description = function_metadata["'description'"]
```
Comment (bkeryan) [inline src/codegen/properties/functions.py:14]: The key name should not contain single quotes.

File: src/codegen/properties/functions.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        self._return_type = function_metadata["returns"]
+        if "handle_parameter" in function_metadata:
+            self._handle_parameters = []
+            for name, parameter_data in function_metadata["handle_parameters"].items():
+                self._handle_parameters.append(Parameter(name, parameter_data))
+
+        if "parameters" in function_metadata:
+            self._parameters = []
+            for parameter in function_metadata["parameters"].items():
+                self._parameters.append(FunctionsParameter(parameter))
```
Comment (bkeryan) [inline src/codegen/properties/functions.py:27]: Isn't `parameters` required? Also, `parameters` is a `list` so why call `.items()`?

FYI, you can write this as a list comprehension:
```
self._parameters = [FunctionParameter(p) for p in function_metadata["parameters"]]
```

File: src/codegen/properties/functions.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        self._c_function_name = function_metadata["c_function_name"]
+        self._description = function_metadata["'description'"]
+        self._is_python_factory = function_metadata.get("is_python_factory", False)
+        self._python_class_name = function_metadata["python_class_name"]
+        self._calling_convention = function_metadata["calling_convention"]
+        self._return_type = function_metadata["returns"]
+        if "handle_parameter" in function_metadata:
+            self._handle_parameters = []
+            for name, parameter_data in function_metadata["handle_parameters"].items():
+                self._handle_parameters.append(Parameter(name, parameter_data))
```
Comment (bkeryan) [inline src/codegen/properties/functions.py:22]: `handle_parameter` is singular. It describes a single parameter. You can't build a list out of it.

```
    'CalculateReversePolyCoeff': {
        'calling_convention': 'StdCall',
        'handle_parameter': {
            'ctypes_data_type': 'ctypes.c_char_p',
            'cvi_name': 'name',
            'python_accessor': 'self._name'
        },
        'parameters': [
            {
                'direction': 'in',
```

File: src/codegen/properties/function_parameters.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
+"""Structure for storing function parameters metadata from scrapigen."""
+from codegen.properties.cluster_elements import ClusterElements
+
+
+class FunctionsParameter:
+    """Structure for storing function metadata from scrapigen."""
+
+    def __init__(self, parameters_metadata):
```
Comment (bkeryan) [inline src/codegen/properties/function_parameters.py:8]: `parameters_metadata` describes a single parameter so rename it to `parameter_metadata`.

File: src/codegen/properties/function_parameters.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
+"""Structure for storing function parameters metadata from scrapigen."""
+from codegen.properties.cluster_elements import ClusterElements
+
+
+class FunctionsParameter:
+    """Structure for storing function metadata from scrapigen."""
```
Comment (bkeryan) [inline src/codegen/properties/function_parameters.py:6]: This class's docstrings are missing the word "parameter".

File: src/codegen/properties/function_parameters.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        self._default = parameters_metadata.get("default", "")
+        self._cluster = parameters_metadata.get("cluster", "")
+        self._is_enum = False
+        if "enum" in parameters_metadata:
+            self._enum = parameters_metadata.get("enum")
+            self._is_enum = True
+        if "cluster_elements" in parameters_metadata:
+            self._cluster_elements = []
+            for cluster_element in parameters_metadata.get("cluster_elements"):
+                self._cluster_elements.append(ClusterElements(cluster_element))
```
Comment (bkeryan) [inline src/codegen/properties/function_parameters.py:28]: `cluster_element` isn't in https://github.com/ni/grpc-device-scrapigen/pull/143 so I don't know if anything related to it is correct.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: src/codegen/properties/functions.py (hunk 2 above)
Comment (Sakthi-SM) [reply src/codegen/properties/functions.py]: Fixed the typos

File: src/codegen/properties/function_parameters.py
[the remark this answers is not in the document]
Comment (Sakthi-SM) [reply src/codegen/properties/function_parameters.py:1]: Updated the file structure as per the suggestion

**Review by zhindes — APPROVED** [review]

File: src/codegen/functions/function.py [hunk 8]
```diff
@@ -66,7 +69,7 @@ def return_type(self):
 
     @property
     def handle_parameters(self):
-        """List of handle parameters: The handle parameter for the instance the function."""
+        """Handle parameter: The handle parameter for the instance the function."""
         return self._handle_parameters
```
Comment (bkeryan) [inline src/codegen/functions/function.py]: Plural -> singular

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: src/codegen/properties/functions.py (hunk 3 above)
Comment (Sakthi-SM) [reply src/codegen/properties/functions.py]: Yes Brad `parameters` is a required field. I have removed the check to see if the field is present

**Review by bkeryan — APPROVED** [review]

**Pull request merged**